Implement split_with_sizes backward for NT

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110647
* #110646

Needed internally. Note that `split_with_sizes()` for NT is currently supported only on `dim=-1`.

jbschlosser committed:
Implement split_with_sizes backward for NT

[ghstack-poisoned]

cpuhrsch reviewed (approved):
Stamp

jbschlosser commented:
@pytorchbot merge

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2116,8 +2116,35 @@ Tensor _nested_split_with_sizes_backward(
+    const Tensor& nt_sizes,
+    const at::TensorOptions& options) {
+  // add 1 to account for batch dim
+  dim = at::maybe_wrap_dim(dim, static_cast<int64_t>(nt_sizes.size(1)) + 1);
```
nit: maybe we don't need the static casts?

kit1980 commented:
I've cancelled the merge as https://github.com/pytorch/pytorch/pull/110404 is likely to be reverted.

jbschlosser committed:
Update on "Implement split_with_sizes backward for NT"


Needed internally. Note that `split_with_sizes()` for NT is currently supported only on `dim=-1`.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
